File: cpp/2533.cpp
#include <iostream>
#include <vector>
#include <cstring>

std::vector<std::vector<int>> tree;
int cache[1000001][2];
int visited[1000001];

int dfs(int i, int ea)
{
	if (tree[i].size() == 0)
	{
		return ea;
	}

	if (cache[i][ea] > -1)
	{
		return cache[i][ea];
	}

	visited[i] = 1;

	cache[i][ea] = ea;
	if (ea)
	{
		for (int j = 0; j < tree[i].size(); ++j)
		{
			if (!visited[tree[i][j]])
			{
				int a = dfs(tree[i][j], 0);
				int b = dfs(tree[i][j], 1);
				cache[i][ea] += a < b ? a : b;
			}
		}
	}
	else
	{
		for (int j = 0; j < tree[i].size(); ++j)
		{
			if (!visited[tree[i][j]])
			{
				cache[i][ea] += dfs(tree[i][j], 1);
			}
		}
	}

	visited[i] = 0;

	return cache[i][ea];
}

int main()
{
	int n;
	std::cin >> n;

	for (int i = 0; i < n; ++i)
	{
		tree.push_back(std::vector<int>());
	}

	memset(cache, -1, sizeof(cache));
	
	for (int i = 0; i < n - 1; ++i)
	{
		int u, v;

		std::cin >> u >> v;
		tree[u - 1].push_back(v - 1);
		tree[v - 1].push_back(u - 1);
	}

	int a = dfs(0, 0);
	int b = dfs(0, 1);

	std::cout << (a < b ? a : b) << std::endl;
}

File: python/2533.py
import sys

sys.setrecursionlimit(100000010)

n = int(input())

tree = [[] for _ in range(n)]

for _ in range(n - 1):
    u, v = map(int, input().split(" "))
    tree[u - 1].append(v - 1)

cache = [[-1, -1] for _ in range(n)]

def dfs(i=0, ea=0):
    if len(tree[i]) == 0:
        return ea

    if cache[i][ea] >= 0:
        return cache[i][ea]

    cache[i][ea] = ea
    if ea:
        for j in tree[i]:
            a = cache[j][0] if cache[j][0] >= 0 else dfs(j, 0)
            b = cache[j][1] if cache[j][1] >= 0 else dfs(j, 1)
            cache[i][ea] += min(a, b)
    else:
        for j in tree[i]:
            cache[i][ea] += cache[j][1] if cache[j][1] >= 0 else dfs(j, 1)

    return cache[i][ea]
print(min(dfs(0, 0), dfs(0, 1)))